### DISCUSSION TITLE
When should we use WithViewStore vs an @ObservedObject viewStore?

### DISCUSSION THREAD

**Discussion by mycroftcanner** [pointfreeco/swift-composable-architecture#856] [Q&A] [answered]
The documentation says there is an issue with GeometryReader but it doesn't mention what are the performance implications and if there are special considerations for each situation.

I keep discovering new little [things](https://forums.swift.org/t/how-do-you-best-pass-scoped-stores-thru-multiple-view-layers/43849)...  
 
In isowords, I see :

```swift
    self.viewStore = ViewStore(self.store, removeDuplicates: ==)
```

When does it make sense to use removeDuplicates?

Has anyone benchmarked using WithViewStore vs @ObservedObject ?

**Comment by stephencelis** [comment, marked as answer]
That `removeDuplicates: ==` can probably go! The initializer that doesn't take that parameter is constrained to `Equatable` and calls `==` under the hood:

https://github.com/pointfreeco/swift-composable-architecture/blob/14379519edb3ddeb95c3262ad3d808d42c146af9/Sources/ComposableArchitecture/SwiftUI/WithViewStore.swift#L125-L140

> When does it make sense to use removeDuplicates?

Whenever you want to perform a custom equatable check, or whenever you're dealing with types that aren't equatable (like tuples).

> Has anyone benchmarked using WithViewStore vs @ObservedObject ?

`WithViewStore` is a view that holds onto an `@ObservedObject` under the hood. The main difference is that `WithViewStore` performs an equatable check (via that `removeDuplicates` parameter) to prevent recomputing the body when state hasn't changed). This means more calls to `==`, and fewer calls to `body`. If `==` is expensive, `WithViewStore` may incur a performance penalty. If `body` is expensive, `WithViewStore` may be more performant by default. You can also pass `removeDuplicates: { _, _ in false }` to `WithViewStore` to opt out of the equatable check, at which point it should be equivalent, functionally and performance-wise.
